### PR TITLE
Typing for permuta/perm.py

### PR DESCRIPTION
**Pull request by JonSteinn**
Typing. Naming standardized for perm length and indices and elements in perms. Some refactoring. In `permuta/perm.py`.

### CODE REVIEW

**Comment by enadeau** [comment]
This looks promising, I added mypy to the test suite so that the typing is checked.

File: permuta/interfaces/flippable.py [hunk 1]
```diff
@@ -1,9 +1,9 @@
 import abc
 
-ABC = abc.ABCMeta("ABC", (object,), {})
+# ABC = abc.ABCMeta("ABC", (object,), {})
```
Comment (enadeau) [inline permuta/interfaces/flippable.py:3]: Probably better to remove then comment

File: permuta/interfaces/patt.py [hunk 2]
```diff
@@ -1,9 +1,10 @@
 import abc
+from typing import Iterator, List, Tuple, Union
 
-ABC = abc.ABCMeta("ABC", (object,), {})
+# ABC: Type = abc.ABCMeta("ABC", (object,), {})
```
Comment (enadeau) [inline permuta/interfaces/patt.py:4]: Also remove comment

File: permuta/interfaces/rotatable.py [hunk 3]
```diff
@@ -1,9 +1,9 @@
 import abc
 
-ABC = abc.ABCMeta("ABC", (object,), {})
+# ABC = abc.ABCMeta("ABC", (object,), {})
```
Comment (enadeau) [inline permuta/interfaces/rotatable.py:3]: Same

File: permuta/interfaces/shiftable.py [hunk 4]
```diff
@@ -1,9 +1,9 @@
 import abc
 
-ABC = abc.ABCMeta("ABC", (object,), {})
+# ABC = abc.ABCMeta("ABC", (object,), {})
```
Comment (enadeau) [inline permuta/interfaces/shiftable.py:3]: Same

File: permuta/perm.py [hunk 5]
```diff
@@ -53,29 +62,31 @@ def __new__(cls, iterable=(), check=False):
         """
         return tuple.__new__(cls, iterable)
 
-    def __init__(self, iterable=(), check=False):
+    def __init__(self, iterable: Iterable = (), check: bool = False) -> None:
```
Comment (enadeau) [inline permuta/perm.py:65]: Would it be better to specify `Iterable[int]`, maybe also for `__new__`

File: permuta/perm.py [hunk 6]
```diff
@@ -53,29 +62,31 @@ def __new__(cls, iterable=(), check=False):
...
+            if used[val]:
+                raise ValueError("Duplicate element: {}".format(val))
+            used[val] = True
 
-    _to_standard_cache = {}
+    _to_standard_cache: ClassVar[Dict[Tuple, "Perm"]] = {}
 
     @classmethod
-    def to_standard(cls, iterable):
+    def to_standard(cls, iterable: Iterable) -> "Perm":
```
Comment (enadeau) [inline permuta/perm.py:89]: `Iterable[int]` ?

File: permuta/perm.py [hunk 7]
```diff
@@ -1247,39 +1103,34 @@ def count_non_inversions(self):
...
             >>> tuple(Perm((3, 0, 2, 1, 4)).non_inversions())
             ((0, 4), (1, 2), (1, 3), (1, 4), (2, 4), (3, 4))
         """
-        for i in range(len(self)):
-            for j in range(i + 1, len(self)):
-                if self[i] < self[j]:
-                    yield (i, j)
+        n = len(self)
+        for i, prev in enumerate(self):
+            for j in range(i + 1, n):
```
Comment (enadeau) [inline permuta/perm.py:1116]: This feels like the good place to use `islice`. Same for the inversions

File: permuta/perm.py [hunk 8]
```diff
@@ -1952,28 +1736,15 @@ def count_occurrences_of(self, patt):
...
-            The indices of the occurrences of self in patt.
-            Each yielded element l is a tuple of integer indices of the
-            pattern patt such that:
-            self == permuta.Perm.to_standard([patt[i] for i in l])
+    def occurrences_in(
+        self,
+        patt: "Patt",
+        self_colours: List[int] = None,
+        patt_colours: List[int] = None,
+    ) -> Union[Iterator[List[Tuple[int]]], Iterator[Tuple[()]]]:
```
Comment (enadeau) [inline permuta/perm.py:1744]: I think this output should only be `Iterator[Tuple[int,...]]]` as the tuple can be of any length depending on of the length of the pattern

File: permuta/perm.py [hunk 9]
```diff
@@ -2069,21 +1840,12 @@ def occurrences(i, k):
...
-            self:
-                A perm.
-            patt: <permuta.interfaces.Patt>
-                A classical/mesh pattern.
-
-        Yields: <tuple> of <int>
-            The indices of the occurrences of self in perm.
+    def occurrences_of(
+        self, patt: "Patt"
+    ) -> Union[Iterator[List[Tuple[int]]], Iterator[Tuple[()]]]:
```
Comment (enadeau) [inline permuta/perm.py:1845]: Same as `occurences_in`

File: permuta/perm.py [hunk 10]
```diff
@@ -2271,63 +2002,48 @@ def __call__(self, value):
...
         """
-        if not isinstance(value, numbers.Integral):
-            raise TypeError("'{}' object is not an integer".format(repr(value)))
-        if not 0 <= value < len(self):
-            raise ValueError("Element out of range: {}".format(value))
+        assert 0 <= value < len(self)
         return self[value]
 
-    def __add__(self, other):
+    def __add__(self, other) -> "Perm":
```
Comment (enadeau) [inline permuta/perm.py:2008]: I think it's more pythonic to check the type of other and `return NotImplemented` when the class of the object is not a Perm

File: permuta/perm.py (hunk 7 above)
Comment (JonSteinn) [reply permuta/perm.py:1116]: `islice` actually goes through the entire iterable and consumes it up to starting index so given `i=n-2`, the second loop would still iterate through the entire perm and just dump those up to `n-1`.

That's why I kept the second range.

File: permuta/perm.py (hunk 6 above)
Comment (JonSteinn) [reply permuta/perm.py:89]: I think this is used from all sort of iterables. The example in the docstring even has one with a string.

File: permuta/perm.py (hunk 5 above)
Comment (JonSteinn) [reply permuta/perm.py:65]: Agreed

File: permuta/perm.py (hunk 7 above)
Comment (enadeau) [reply permuta/perm.py:1116]: Ok make sense. I learned something

File: permuta/perm.py (hunk 6 above)
Comment (enadeau) [reply permuta/perm.py:89]: Seems reasonable even if in practice we only ever use int

**Pull request merged**